src/js/components/Button/Button.js

```diff
@@ -343,9 +370,87 @@ const Button = forwardRef(
     }
     if (tip) {
       if (typeof tip === 'string') {
-        return <Tip content={tip}>{styledButtonResult}</Tip>;
+        styledButtonResult = <Tip content={tip}>{styledButtonResult}</Tip>;
       }
-      return <Tip {...tip}>{styledButtonResult}</Tip>;
+      styledButtonResult = <Tip {...tip}>{styledButtonResult}</Tip>;
+    }
+
+    const defaultBadgeDimension =
+      typeof badgeProp === 'boolean'
+        ? // empty badge should be smaller
+          `${parseInt(theme.button.badge.size.medium.replace('px', ''), 10) /
+            2}px`
+        : theme.button.badge.size.medium;
+
+    const [badgeWidth, setBadgeWidth] = useState(defaultBadgeDimension);
+    // if badge has a number, we want to scale the badge to fit.
+    // width will also be used to know how far to horizontally offset the badge
+    useEffect(() => {
+      if (badgeRef && badgeRef.current && typeof badgeProp !== 'boolean') {
+        setBadgeWidth(getDimension('width', badgeProp, badgeRef, theme));
+      }
+    }, [badgeProp, theme]);
+
+    const [badgeHeight, setBadgeHeight] = useState(defaultBadgeDimension);
+    // if badge has custom JSX, we want to scale the badge to fit.
+    // height will also be used to know how far to horizontally offset the badge
+    useEffect(() => {
+      if (badgeRef && badgeRef.current && typeof badgeProp !== 'boolean')
+        setBadgeHeight(getDimension('height', badgeProp, badgeRef, theme));
+    }, [badgeProp, theme]);
+
+    if (badgeProp) {
+      const max = badgeProp.max || theme.button.badge.max;
+
+      let value;
+      if (typeof badgeProp === 'number') value = badgeProp;
+      else if (typeof badgeProp === 'object') value = badgeProp.value;
+
+      let badge;
+      if (value || typeof badgeProp === 'boolean') {
+        if (value) {
+          badge = (
+            <Text color="text-strong" size="small" ref={badgeRef}>
+              {value > max ? `${max}+` : value}
```

Comment:
Are we okay here if max is not defined by neither the badgeProp nor the theme?

Reply:
Yes, we check if there is a value before comparing to max. If so, value > undefined will resolve to false

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Greater_than#comparing_boolean_null_undefined_nan